module/Auth/src/Adapter/LdapAdapter.php

```diff
@@ -0,0 +1,212 @@
+     */
+    public function register(string $identifier, string $password, string $email, array $attributes = []): void
+    {
+        $attributes = array_merge(['email' => $email], $attributes);
+        $this->client->register($identifier, $password, $attributes);
+    }
+
+    public function changeExpiredPassword(string $newPassword, string $challengeToken, string $username): ChangeExpiredPasswordResult
+    {
+        try {
+            $token = $this->client->authenticate($username, $newPassword);
+        } catch (ClientException $e) {
+            Logger::err('Ldap client: change password ClientException checking previous password: ' . $e->getMessage());
+            return new ChangeExpiredPasswordResult(ChangeExpiredPasswordResult::FAILURE_CLIENT_ERROR, [], [$e->getMessage()]);
+        }
+
+        return new ChangeExpiredPasswordResult(ChangeExpiredPasswordResult::SUCCESS, $this->buildUserObject($token));
+    }
+
+    public function refreshToken(string $refreshToken, string $identifier): Result
+    {
+        // No refresh token functionality in LDAP.
+        return new Result(Result::FAILURE, []);
+    }
+
+    /**
+     * @throws ClientException
+     */
+    public function changeAttribute(string $identifier, string $key, string $value): void
+    {
+        $this->client->changeAttribute($identifier, $key, $value);
+    }
+
+    public function deleteUser(string $identifier): DeleteUserResult
+    {
+        throw new \RuntimeException('Not implemented');
+    }
+
+    /**
+     * @throws ClientException
+     */
+    public function disableUser(string $identifier): void
+    {
+        $this->client->disableUser($identifier);
+    }
+
+    /**
+     * @throws ClientException
+     */
+    public function enableUser(string $identifier): void
+    {
+        $this->client->enableUser($identifier);
+    }
+
+    /**
+     * @throws ClientException
+     */
+    public function getUserByIdentifier(string $identifier): ResourceOwnerInterface
+    {
+        return $this->client->getUserByIdentifier($identifier);
+    }
+
+    /**
+     * @throws ClientException
+     */
+    public function registerIfNotPresent(string $identifier, string $password, string $email, array $attributes = []): bool
+    {
+        if (!$this->doesUserExist($identifier)) {
+            $this->register($identifier, $password, $email, $attributes);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @throws ClientException
+     */
+    public function doesUserExist(string $identifier): bool
+    {
+        try {
+            $this->getUserByIdentifier($identifier);
+        } catch (ClientException $e) {
+            throw $e;
+        }
+
+        return true;
+    }
+
+    /**
+     * @return mixed|string
+     */
+    public function getIdentity()
+    {
+        $identity = parent::getIdentity();
+        if (!empty($identity) && is_string($identity)) {
+            $identity = strtolower($identity);
```

Comment:
Why are we lowercasing this?

Reply:
It's from the `module/Auth/src/Adapter/CognitoAdapter.php`. I didn't want to tempt fate and remove the logic incase I uncover a world of pain.